blog/_posts/2017-02-25-scalafix-v0.3.md

```diff
@@ -0,0 +1,154 @@
+Semantic databases from different compilation units, potentially produced by different
+versions of the Scala compiler, can be merged.
+This opens possibilities for large-scale code analysis.
+
+The introduction of the scala.meta semantic API is a game changer for scalafix.
+The ability to resolve names to symbols opens possibilities for many scalafix rewrites.
+
+## Rewrite: meta.Tree => Seq[Patch]
+
+A scalafix `Patch` is a small operation that can produce a diff on a Scala source file.
+A patch can either be a "token patch" or a "tree patch".
+
+Token patches are low-level but give full control over how every detail in a source file is handled, for example formatting and comments.
+Example token patches are `Remove(token)` and `AddLeft(token, toAdd: String)`, which removes or prepends a string to `token`, respectively.
+
+Tree patches are high-level and allow rewrite author to declaratively explain what operation to perform.
+An example tree patch is `AddGlobalImport(importer)`, which adds a new import to the top of a file if it does not exist.
+Observe that `AddGlobalImport` does not worry about token-level details such as whether the user groups imports by prefix (e.g., `import a.{b, c}`).
+
+Tree and token patches build a small algebra of operations that can be composed to build complex refactorings.
+A challenge with composing patches is to figure out what to do on conflicts.
+For example, what happens when one patch renames a token while the other patch removes the same token?
+The current strategy in scalafix is to try and resolve as many conflicts as possible on the tree patch level.
+It can be harder to resolve conflicts on the token level since the original intent of the patch is lost.
+Unsolvable conflicts abort the refactoring.
+In the future, we hope to support more advanced conflict resolution strategies.
+
+In a nutshell, a scalafix `Rewrite` is a `scala.meta.Tree => Seq[Patch]` function.
+The tree is backed by the scala.meta semantic API, so the rewrite is able to query for compiler information such as symbols.
+To demonstrate how rewrites are implemented with scalafix v0.3, let's step through an example use-case.
+
+## Example: Xor to Either
+
+The functional programming library [cats][] migrated recently from its `Xor` data type to `Either` from the standard library.
+It requires a few mechanical steps to migrate code to use `Either` instead of `Xor`.
+For example, below is a diff that's taken from [circe][]'s migration to `Either`.
+
+```diff
+-final def either: Xor[HCursor, HCursor] = if (succeeded) Xor.right(any) else Xor.left(any)
++final def either: Either[HCursor, HCursor] = if (succeeded) Right(any) else Left(any)
+```
+
+For this particular rewrite, we are able to get away with only using tree patches.
+`Replace` is one tree patch that can be used to replace usage of a reference such as a type, term or a static method.
+
+```scala
+Replace(Symbol("_root_.cats.data.Xor."), q"Either")
```

Comment:
Trailing dot is required, it indicates a `Signature.{Term,Method}`, `Signature.Type` ends with `#` (e.g., "_root_.scala.util.Failure#").
Currently a missing `.` in the suffix results in a runtime error, but we'll probably make `symbol"_root_.foo."` macro interpolator that will work similar to quasiquotes to make this constructor safe.